data/collector.go

```diff
@@ -178,10 +183,27 @@ func (c *collector) tickerBasedCapture(started chan struct{}) {
 	}
 }
 
+func (c *collector) validateReadingType(t CaptureType) error {
```

Comment:
At a high level, why do both the collector and the reading it returns have a CaptureType? It seems we never expect them to be different, so why do we encode this information in two different places?

Reply:
Great question!

The definition of that is a tabular or binary collector lives [here](https://github.com/viamrobotics/rdk/pull/4557/files#diff-67c0823aa24d3667ac603603cf934d7c0d988547e19a88cd2d48b3aa1337dc99R193) and is called [here](https://github.com/viamrobotics/rdk/pull/4557/files#diff-943a1876bce4656fb8ce485d913b26670b3449d0d430a98002fe1c2439373975R220) which is called [here](https://github.com/viamrobotics/rdk/pull/4557/files#diff-b38086d606e0e0fd934d9795d68af0065d61bfd9b5ea4a960aa69ba9f511615cR327) in capture.go.

That means that (as is also true on main) datamanager/capture/capture.go is responsible for determining whether a given resource/method pair is binary or tabular. 
 
However [data.CaptureFunc](https://github.com/viamrobotics/rdk/pull/4557/files/f3beccf39a296f5d51c4759a83b3dfa1dd59b336#diff-67c0823aa24d3667ac603603cf934d7c0d988547e19a88cd2d48b3aa1337dc99R22) (the interface method that a collector implements) returns a CaptureResult, which can either be binary or tabular. validateReadingType checks that the response (what the collector actually returned) matches the expectations of collector.go.

This is a risk that exists today in main, I'm just now checking for the mismatch.

The only way I can see to resolve this runtime check is to create 2 different constructors collectors can implement, one which returns a binary response and another that returns a tabular response. That would be cleaner, but would take more time for me to implement.

Let me know if that makes sense.

Reply:
Great explanation, thank you so much! I'm happy with the choice you've made to just add validation for now. I like your idea for how to catch this at compile time, we can always do that in the future. This PR is pretty big already :)

Reply:
https://github.com/viamrobotics/rdk/blob/main/services/datamanager/builtin/capture/capture.go#L327
which calls https://github.com/viamrobotics/rdk/blob/main/data/capture_file.go#L220
which calls https://github.com/viamrobotics/rdk/blob/main/data/capture_file.go#L244


The response of data.BuildCaptureMetadata is passed to the collector constructor here: https://github.com/viamrobotics/rdk/blob/main/services/datamanager/builtin/capture/capture.go#L345
which determines the datasync.DataType of the entire capture file: https://github.com/viamrobotics/api/blob/main/proto/viam/app/datasync/v1/data_sync.proto#L134

Nothing in main enforces that the DataCaptureMetadata.Type must match the type of the resource/method pair response.

Reply:
Thanks for linking all of that! LGTM!